shotover/src/transforms/kafka/sink_cluster/mod.rs

```diff
@@ -2860,3 +2925,50 @@ impl<'a> std::fmt::Display for FormatTopicName<'a> {
         }
     }
 }
+
+/// Partitions the given shotover nodes based on their rack placement.
+/// Returns a tuple of two vectors: (shotover_nodes_in_rack, shotover_nodes_out_of_rack).
+/// If rack is None, all shotover nodes are considered to be in the rack.
+fn partition_shotover_nodes_by_rack<'a>(
+    shotover_nodes: &[&'a ShotoverNode],
+    rack: &Option<&StrBytes>,
+) -> (Vec<&'a ShotoverNode>, Vec<&'a ShotoverNode>) {
+    shotover_nodes
+        .iter()
+        .partition(|shotover_node| rack.map(|rack| rack == &shotover_node.rack).unwrap_or(true))
+}
+
+/// Tries to deterministically select a shotover node based on a hash value.
+/// Prioritizes nodes in the rack if available, falls back to nodes out of the rack.
+fn select_shotover_node_by_hash<'a>(
+    shotover_nodes_in_rack: Vec<&'a ShotoverNode>,
+    shotover_nodes_out_of_rack: Vec<&'a ShotoverNode>,
+    hash: usize,
+) -> Option<&'a ShotoverNode> {
+    if !shotover_nodes_in_rack.is_empty() {
+        shotover_nodes_in_rack
+            .get(hash % shotover_nodes_in_rack.len())
+            .cloned()
+    } else if !shotover_nodes_out_of_rack.is_empty() {
+        shotover_nodes_out_of_rack
+            .get(hash % shotover_nodes_out_of_rack.len())
+            .cloned()
+    } else {
+        None
+    }
+}
+
+/// Returns a set of unique broker IDs from available shotover nodes.
+/// Prioritizes nodes in the rack if available, falls back to nodes out of the rack.
+fn collect_unique_broker_ids(
+    shotover_nodes_in_rack: Vec<&ShotoverNode>,
+    shotover_nodes_out_of_rack: Vec<&ShotoverNode>,
+) -> HashSet<BrokerId> {
```

Comment:
Instead of converting to a hashset which requires extra allocation and hashing (expensive) we should have an assertion on startup that each ShotoverNode has a unique BrokerId

Reply:
makes sense. will do

Reply:
added the check